src/kalai/pass/kalai/a_annotate_ast.clj

```diff
@@ -54,67 +54,55 @@
 
 (defn ast-t
   "Return the type represented by an AST node"
-  [ast]
-  (m/rewrite ast
-    ;; (atom x)
-    {:op   :invoke
-     :fn   {:var (m/pred ref-vars)}
-     :args [?value . _ ...]}
-    ~(ast-t ?value)
-
-    ;; ^{:t :long, :tag Long} x
-    {:op   :with-meta
-     :meta {:form {:t ?t :tag ?tag}}
-     :expr ?expr}
-    ~(or ?t
-         (get types/java-types ?tag)
-         (ast-t ?expr))
-
-    {:op :binding
-     #_#_:meta {:form {:t ?t :tag ?tag}
-                :as ?meta}
-     :form (m/and
-             (m/app #(resolve-t % ast) ?t)
-             (m/guard (some? ?t)))
-     :as ?binding}
-    ~(do (u/spy [:t ?t] ":binding with :meta")
-         (u/spy (-> ?binding :form meta) "?binding form meta")
-         ?t)
-
-    {:op :binding
-     :init ?init
-     :as ?binding}
-    ~(do (u/spy ?binding ":binding without :meta")
-         (println (-> ?binding :form meta))
-         (println (-> ?binding :name meta))
-         (ast-t ?init))
-
-    {:op :binding
-     :init ?init}
-    ~(do (u/spy (ast-t ?init) ":binding without :meta")
-       (ast-t ?init))
-
-    {:op   :local
-     :form ?form
-     :env {:locals {?form ?binding}}}
-    ~(ast-t ?binding)
-
-    ;; Last resort: 1
-    {:op :const
-     :val ?val}
-    ~(get types/java-types (type ?val))
-
-    ;; Last resort: Clojure type inferred
-    {:o-tag (m/pred some? ?o-tag)}
-    ~(or (get types/java-types ?o-tag)
-         ;; TODO: breaking the rules for interop...
-         ;; is this a bad idea?
-         ;; note that some o-tags are unhelpful,
-         ;; like PersistentMap
-         (#{StringBuffer} ?o-tag))
-
-    ?else
-    nil))
+  ([ast]
+   (ast-t ast ast))
+  ([ast root]
+   (m/rewrite ast
+     ;; (atom x)
+     {:op   :invoke
+      :fn   {:var (m/pred ref-vars)}
+      :args [?value . _ ...]}
+     ~(ast-t ?value ast)
+
+     ;; ^{:t :long, :tag Long} x
+     {:op   :with-meta
+      :meta {:form {:t ?t :tag ?tag}}
+      :expr ?expr}
+     ~(or ?t
+          (get types/java-types ?tag)
+          (ast-t ?expr ast))
+
+     {:op   :binding
+      :form (m/and
+              (m/app #(resolve-t % root) ?t)
```

Comment:
This commit fixed the test breakage. What I'm puzzled by is how the extra function arity that also adds the argument `root` (in addition to `ast`) and its usage here made the difference. The reason I wouldn't expect a change is that I don't see any new references to the 2-arity version of `ast-t`, except from the 1-arity version's body, and it just passes the value of `ast` to be the value of `root` anyways.

Reply:
All of the recursive calls to `ast-t` are now the 2-arity call. Your IDE might just be confused about meander macros and doesn't find them. The philosophy here is we always call `ast-t` on a valid node that has a ns in the environment. But when we recur down into locals, we get "nodes" that do not... but we can just use the root ns for resolving as it must be the same.

Reply:
Ahh, that makes sense, I get it now.